File: 397__Integer_Replacement.py
class Solution(object):
    def integerReplacement(self, n):
        """
        :type n: int
        :rtype: int
        """
        result = 0
        while n != 1:
            if bin(n)[-1] == "0":
                n = n >> 1
            else:
                if bin(n)[-2] == "1" and len(bin(n)) != 4:
                    n += 1 # if "11" +1
                else:
                    n -= 1 # if "01" -1
            result += 1
        return result

File: Java/397__Integer_Replacement.java
class Solution {
    public int integerReplacement(int n) {
        int result = 0;
        while (n != 1){
            if ((n & 1) == 0) n = n >> 1;
            else {
                if ((n & 2) == 2) {
                    if (n == 3) {
                        result += 2;
                        return result;
                    }
                    n = n >> 1;
                    n++;
                    result++;
                } else
                    n--;
            }
            result++;
        }
        return result;
    }
}
